File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits

handshake :: Int -> [String]
handshake = process 0 []

process :: Int -> [String] -> Int -> [String]
process num acc n
    | num == 4  = if bitset then acc else reverse acc
    | bitset    = process (num + 1) ((signal !! num):acc) n
    | otherwise = process (num + 1) acc n
    where
        bitset = shiftR n num .&. 1 == 1

signal :: [String]
signal = ["wink", "double blink", "close your eyes", "jump"]

File: typescript/secret-handshake/secret-handshake.ts
const SECRET_HANDSHAKE_OPERATIONS = {
  0b0001: "wink",
  0b0010: "double blink",
  0b0100: "close your eyes",
  0b1000: "jump",
};

const REVERSE_HANDSHAKE_KEY = 0b10000;

export default class Handshake {
  constructor(private readonly secret: number) {}

  commands(): string[] {
    const shake = Object.entries(SECRET_HANDSHAKE_OPERATIONS)
      .sort()
      .filter(([handshakeType, _]) => this.secret & Number(handshakeType))
      .map(([_, handShakeOperation]) => handShakeOperation);

    return this.secret & REVERSE_HANDSHAKE_KEY ? shake.reverse() : shake;
  }
}
